nntrainer/dataset/data_producers.h

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   data_producers.h
+ * @date   09 July 2021
+ * @brief  This file contains data producer interface
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#ifndef __DATA_PRODUCERS_H__
+#define __DATA_PRODUCERS_H__
+
+#include <functional>
+#include <limits>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include <tensor.h>
+#include <tensor_dim.h>
+
+namespace nntrainer {
+
+/**
+ * @brief DataProducer interface used to abstract data provider
+ *
+ */
+class DataProducer {
+public:
+  using Iteration = std::tuple<bool, std::vector<Tensor>, std::vector<Tensor>>;
+
+  /**
+   * @brief create an iteration
+   * @return Iteration iteration, if std::get<0>(retval) == true means end of
+   * iteration, at the end of the iteration, it's responsibility of @a this to
+   * shuffle.
+   */
+  using Gernerator = std::function<Iteration(void)>;
+
+  constexpr inline static unsigned long long SIZE_UNDEFINED =
```

Comment:
how about `size_t`?

Reply:
any reasons? I chose to use unsigned long long as the value is predictable but size_t is not.

Reply:
size_t is standard to store the size of the dataset and scales its byte size appropriately as per systems. For example, most the STL data structures in C++ use size_t to give out size.